File: src/backend/Intelibill.Domain/Entities/SaleReturn.cs
using ErrorOr;
using Intelibill.Domain.Common;
using Intelibill.Domain.Enums;
using Intelibill.Domain.ValueObjects;

namespace Intelibill.Domain.Entities;

public sealed class SaleReturn : BaseEntity
{
    private readonly List<SaleReturnItem> _items = [];

    public Guid ShopId { get; private set; }
    public Guid SaleId { get; private set; }
    public string ReturnNumber { get; private set; } = string.Empty;
    public DateTimeOffset ProcessedAt { get; private set; }
    public Guid ProcessedBy { get; private set; }
    public string? Notes { get; private set; }
    public decimal TotalRefundAmount { get; private set; }
    public decimal DueReductionAmount { get; private set; }
    public decimal PayoutAmount { get; private set; }
    public decimal TotalTaxableAmount { get; private set; }
    public decimal TotalTaxAmount { get; private set; }
    public decimal? CustomerBalanceBefore { get; private set; }
    public decimal? CustomerBalanceAfter { get; private set; }
    public bool IsVoided { get; private set; }
    public DateTimeOffset? VoidedAt { get; private set; }
    public Guid? VoidedBy { get; private set; }
    public string? VoidReason { get; private set; }

    public IReadOnlyList<SaleReturnItem> Items => _items.AsReadOnly();

    private SaleReturn() { }

    public static ErrorOr<SaleReturn> Record(
        Guid shopId,
        Guid saleId,
        string returnNumber,
        DateTimeOffset processedAt,
        Guid processedBy,
        string? notes,
        decimal totalRefundAmount,
        decimal dueReductionAmount,
        decimal payoutAmount,
        PaymentMethod? payoutMethod,
        decimal totalTaxableAmount,
        decimal totalTaxAmount,
        decimal? customerBalanceBefore,
        decimal? customerBalanceAfter,
        IReadOnlyList<SaleReturnLineInput> lines)
    {
        if (string.IsNullOrWhiteSpace(returnNumber))
        {
            return Error.Validation("SaleReturn.ReturnNumberRequired", "Return number is required.");
        }

        if (totalRefundAmount < 0 || dueReductionAmount < 0 || payoutAmount < 0 || totalTaxableAmount < 0 || totalTaxAmount < 0)
        {
            return Error.Validation("SaleReturn.AmountNegative", "Return amounts cannot be negative.");
        }

        if (dueReductionAmount + payoutAmount != totalRefundAmount)
        {
            return Error.Validation("SaleReturn.RefundSplitMismatch", "Due reduction plus payout must equal total refund.");
        }

        var noteValidation = ValidateLineNotes(lines);
        if (noteValidation.IsError)
        {
            return noteValidation.Errors;
        }

        var payoutValidation = ValidatePayoutMethod(payoutAmount, payoutMethod);
        if (payoutValidation.IsError)
        {
            return payoutValidation.Errors;
        }

        var items = new List<SaleReturnItem>(lines.Count);
        foreach (var line in lines)
        {
            var itemResult = SaleReturnItem.Create(
                line.ShopId,
                saleId,
                line.SaleItemId,
                line.Quantity,
                line.Condition,
                line.OriginalCostPrice,
                line.OriginalSalesPrice,
                line.OriginalTaxRatePercent,
                line.OriginalIsPriceIncludingTax,
                line.MaxRefundAmount,
                line.ApprovedRefundAmount,
                line.TaxableAmount,
                line.TaxAmount,
                NormalizeOptional(line.Notes));

            if (itemResult.IsError)
            {
                return itemResult.Errors;
            }

            items.Add(itemResult.Value);
        }

        var saleReturn = new SaleReturn
        {
            ShopId = shopId,
            SaleId = saleId,
            ReturnNumber = returnNumber.Trim(),
            ProcessedAt = processedAt,
            ProcessedBy = processedBy,
            Notes = NormalizeOptional(notes),
            TotalRefundAmount = totalRefundAmount,
            DueReductionAmount = dueReductionAmount,
            PayoutAmount = payoutAmount,
            TotalTaxableAmount = totalTaxableAmount,
            TotalTaxAmount = totalTaxAmount,
            CustomerBalanceBefore = customerBalanceBefore,
            CustomerBalanceAfter = customerBalanceAfter,
            IsVoided = false,
        };

        saleReturn._items.AddRange(items);
        return saleReturn;
    }

    public ErrorOr<Success> Void(DateTimeOffset voidedAt, Guid voidedBy, string reason)
    {
        if (IsVoided)
        {
            return Error.Validation("SaleReturn.AlreadyVoided", "Sale return is already voided.");
        }

        if (string.IsNullOrWhiteSpace(reason))
        {
            return Error.Validation("SaleReturn.VoidReasonRequired", "Void reason is required.");
        }

        IsVoided = true;
        VoidedAt = voidedAt;
        VoidedBy = voidedBy;
        VoidReason = reason.Trim();

        return Result.Success;
    }

    private static ErrorOr<Success> ValidateLineNotes(IReadOnlyList<SaleReturnLineInput> lines)
    {
        foreach (var line in lines)
        {
            if (!string.IsNullOrWhiteSpace(line.Notes))
            {
                continue;
            }

            if (line.Condition == SaleReturnCondition.Wastage)
            {
                return Errors.Sale.ReturnNoteRequired("wastage returns");
            }

            if (line.ApprovedRefundAmount == 0m)
            {
                return Errors.Sale.ReturnNoteRequired("zero refunds");
            }

            if (line.ApprovedRefundAmount < line.MaxRefundAmount)
            {
                return Errors.Sale.ReturnNoteRequired("partial refunds");
            }
        }

        return Result.Success;
    }

    private static ErrorOr<Success> ValidatePayoutMethod(decimal payoutAmount, PaymentMethod? payoutMethod)
    {
        if (payoutAmount <= 0m)
        {
            return Result.Success;
        }

        if (!payoutMethod.HasValue)
        {
            return Errors.Sale.ReturnPayoutMethodRequired;
        }

        return payoutMethod.Value is PaymentMethod.Cash or PaymentMethod.UPI or PaymentMethod.Card
            ? Result.Success
            : Errors.Sale.ReturnPayoutMethodInvalid;
    }

    private static string? NormalizeOptional(string? value) =>
        string.IsNullOrWhiteSpace(value) ? null : value.Trim();
}

File: src/mobile/android/intelibill_mobile/lib/src/features/sales/domain/entities/sale_return.dart
import 'package:equatable/equatable.dart';

class SaleReturnLineDraft extends Equatable {
  const SaleReturnLineDraft({
    required this.saleItemId,
    required this.selected,
    required this.quantity,
    this.lineType,
    this.condition,
    this.approvedRefundAmount,
    this.notes,
  });

  final String saleItemId;
  final bool selected;
  final double quantity;
  final String? lineType;
  final int? condition;
  final double? approvedRefundAmount;
  final String? notes;

  SaleReturnLineDraft copyWith({
    bool? selected,
    double? quantity,
    String? lineType,
    int? condition,
    double? approvedRefundAmount,
    String? notes,
    bool clearLineType = false,
    bool clearCondition = false,
    bool clearApprovedRefundAmount = false,
    bool clearNotes = false,
  }) {
    return SaleReturnLineDraft(
      saleItemId: saleItemId,
      selected: selected ?? this.selected,
      quantity: quantity ?? this.quantity,
      lineType: clearLineType ? null : (lineType ?? this.lineType),
      condition: clearCondition ? null : (condition ?? this.condition),
      approvedRefundAmount: clearApprovedRefundAmount
          ? null
          : (approvedRefundAmount ?? this.approvedRefundAmount),
      notes: clearNotes ? null : (notes ?? this.notes),
    );
  }

  Map<String, dynamic> toRequestMap() {
    return {
      'saleItemId': saleItemId,
      if (lineType != null) 'lineType': lineType,
      'quantity': quantity,
      if (condition != null) 'condition': condition,
      if (approvedRefundAmount != null)
        'approvedRefundAmount': approvedRefundAmount,
      if (notes != null && notes!.trim().isNotEmpty) 'notes': notes,
    };
  }

  @override
  List<Object?> get props => [
    saleItemId,
    selected,
    quantity,
    lineType,
    condition,
    approvedRefundAmount,
    notes,
  ];
}

class SaleReturnPreviewLineFinancial extends Equatable {
  const SaleReturnPreviewLineFinancial({
    required this.originalCostPrice,
    required this.originalSalesPrice,
    required this.originalTaxRatePercent,
    required this.originalIsPriceIncludingTax,
    required this.maxRefundAmount,
    required this.approvedRefundAmount,
    required this.taxableAmount,
    required this.taxAmount,
  });

  final double originalCostPrice;
  final double originalSalesPrice;
  final double originalTaxRatePercent;
  final bool originalIsPriceIncludingTax;
  final double maxRefundAmount;
  final double approvedRefundAmount;
  final double taxableAmount;
  final double taxAmount;

  @override
  List<Object?> get props => [
    originalCostPrice,
    originalSalesPrice,
    originalTaxRatePercent,
    originalIsPriceIncludingTax,
    maxRefundAmount,
    approvedRefundAmount,
    taxableAmount,
    taxAmount,
  ];
}

class SaleReturnPreviewLine extends Equatable {
  const SaleReturnPreviewLine({
    required this.saleItemId,
    this.itemId,
    this.inventoryBatchId,
    required this.requestedQuantity,
    required this.returnedQuantity,
    required this.returnableQuantity,
    this.condition,
    required this.willRestock,
    this.financial,
  });

  final String saleItemId;
  final String? itemId;
  final String? inventoryBatchId;
  final double requestedQuantity;
  final double returnedQuantity;
  final double returnableQuantity;
  final int? condition;
  final bool willRestock;
  final SaleReturnPreviewLineFinancial? financial;

  @override
  List<Object?> get props => [
    saleItemId,
    itemId,
    inventoryBatchId,
    requestedQuantity,
    returnedQuantity,
    returnableQuantity,
    condition,
    willRestock,
    financial,
  ];
}

class SaleReturnPreviewFinancial extends Equatable {
  const SaleReturnPreviewFinancial({
    required this.totalRefundAmount,
    required this.dueReductionAmount,
    required this.payoutAmount,
    required this.totalTaxableAmount,
    required this.totalTaxAmount,
    this.customerBalanceBefore,
    this.customerBalanceAfter,
  });

  final double totalRefundAmount;
  final double dueReductionAmount;
  final double payoutAmount;
  final double totalTaxableAmount;
  final double totalTaxAmount;
  final double? customerBalanceBefore;
  final double? customerBalanceAfter;

  @override
  List<Object?> get props => [
    totalRefundAmount,
    dueReductionAmount,
    payoutAmount,
    totalTaxableAmount,
    totalTaxAmount,
    customerBalanceBefore,
    customerBalanceAfter,
  ];
}

class SaleReturnPreview extends Equatable {
  const SaleReturnPreview({
    required this.saleId,
    required this.hasFinancialAccess,
    required this.lines,
    this.financial,
    required this.warnings,
  });

  final String saleId;
  final bool hasFinancialAccess;
  final List<SaleReturnPreviewLine> lines;
  final SaleReturnPreviewFinancial? financial;
  final List<String> warnings;

  @override
  List<Object?> get props => [
    saleId,
    hasFinancialAccess,
    lines,
    financial,
    warnings,
  ];
}

class PreviewSaleReturnRequest {
  const PreviewSaleReturnRequest({
    this.dueReductionOverrideAmount,
    this.dueOverrideReason,
    required this.items,
  });

  final double? dueReductionOverrideAmount;
  final String? dueOverrideReason;
  final List<SaleReturnLineDraft> items;

  Map<String, dynamic> toJson() {
    return {
      if (dueReductionOverrideAmount != null)
        'dueReductionOverrideAmount': dueReductionOverrideAmount,
      if (dueOverrideReason != null) 'dueOverrideReason': dueOverrideReason,
      'items': items.map((item) => item.toRequestMap()).toList(),
    };
  }
}

class RecordSaleReturnRequest {
  const RecordSaleReturnRequest({
    this.payoutDestination,
    this.payoutMethod,
    this.dueReductionOverrideAmount,
    this.dueOverrideReason,
    this.notes,
    this.creditNoteExpiresAt,
    this.creditNoteReason,
    required this.items,
  });

  final int? payoutDestination;
  final int? payoutMethod;
  final double? dueReductionOverrideAmount;
  final String? dueOverrideReason;
  final String? notes;
  final String? creditNoteExpiresAt;
  final String? creditNoteReason;
  final List<SaleReturnLineDraft> items;

  Map<String, dynamic> toJson() {
    return {
      if (payoutDestination != null) 'payoutDestination': payoutDestination,
      if (payoutMethod != null) 'payoutMethod': payoutMethod,
      if (dueReductionOverrideAmount != null)
        'dueReductionOverrideAmount': dueReductionOverrideAmount,
      if (dueOverrideReason != null) 'dueOverrideReason': dueOverrideReason,
      if (notes != null) 'notes': notes,
      if (creditNoteExpiresAt != null)
        'creditNoteExpiresAt': creditNoteExpiresAt,
      if (creditNoteReason != null) 'creditNoteReason': creditNoteReason,
      'items': items.map((item) => item.toRequestMap()).toList(),
    };
  }
}
